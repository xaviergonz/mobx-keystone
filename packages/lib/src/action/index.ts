export { ActionCall, applyAction } from "./applyAction"
export {
  ActionContext,
  ActionContextActionType,
  ActionContextAsyncStepType,
  getCurrentActionContext,
} from "./context"
export { ActionMiddleware, ActionMiddlewareDisposer, addActionMiddleware } from "./middleware"
export { isModelAction, modelAction } from "./modelAction"
export { FlowRet, isModelFlow, modelFlow } from "./modelFlow"
export { runUnprotected } from "./protection"
export { isHookAction, isSpecialAction, SpecialAction } from "./specialActions"
