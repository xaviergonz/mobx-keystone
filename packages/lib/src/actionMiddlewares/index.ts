export {
  ActionTrackingMiddleware,
  actionTrackingMiddleware,
  ActionTrackingResult,
  ActionTrackingReturn,
  SimpleActionContext,
  simplifyActionContext,
} from "./actionTrackingMiddleware"
export {
  deserializeActionCall,
  deserializeActionCallArgument,
  onActionMiddleware,
  serializeActionCall,
  serializeActionCallArgument,
} from "./onActionMiddleware"
export { readonlyMiddleware, ReadonlyMiddlewareReturn } from "./readonlyMiddleware"
export { transaction, transactionMiddleware } from "./transactionMiddleware"
export { UndoEvent, UndoManager, undoMiddleware, UndoStore, withoutUndo } from "./undoMiddleware"
