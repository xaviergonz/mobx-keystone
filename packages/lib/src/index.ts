export * from "./action"
export * from "./actionMiddlewares"
export * from "./frozen"
export * from "./model"
export * from "./parent"
export * from "./patch"
export * from "./redux"
export * from "./ref"
export * from "./rootStore"
export * from "./snapshot"
export * from "./tweaker"
export { MobxDataModelError } from "./utils"
