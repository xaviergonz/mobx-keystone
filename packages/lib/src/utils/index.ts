import * as mobx from "mobx"
import {
  IObservableArray,
  isObservableArray,
  isObservableMap,
  isObservableObject,
  isObservableSet,
  ObservableMap,
  ObservableSet,
} from "mobx"
import { NonEmptyArray, PrimitiveValue } from "./types"

/**
 * A mobx-keystone error.
 */
export class MobxKeystoneError extends Error {
  constructor(msg: string) {
    super(msg)

    // Set the prototype explicitly.
    Object.setPrototypeOf(this, MobxKeystoneError.prototype)
  }
}

/**
 * @ignore
 * @internal
 */
export function failure(msg: string) {
  return new MobxKeystoneError(msg)
}

const writableHiddenPropDescriptor: PropertyDescriptor = {
  enumerable: false,
  writable: true,
  configurable: false,
  value: undefined,
}

/**
 * @ignore
 * @internal
 */
export function addHiddenProp(object: any, propName: PropertyKey, value: any, writable = true) {
  if (writable) {
    Object.defineProperty(object, propName, writableHiddenPropDescriptor)
    object[propName] = value
  } else {
    Object.defineProperty(object, propName, {
      enumerable: false,
      writable,
      configurable: true,
      value,
    })
  }
}

/**
 * @ignore
 * @internal
 */
export function makePropReadonly<T>(object: T, propName: keyof T, enumerable: boolean) {
  const propDesc = Object.getOwnPropertyDescriptor(object, propName)
  if (propDesc) {
    propDesc.enumerable = enumerable
    if (propDesc.get) {
      delete propDesc.set
    } else {
      propDesc.writable = false
    }
    Object.defineProperty(object, propName, propDesc)
  }
}

/**
 * @ignore
 * @internal
 */
export function isPlainObject(value: any): value is Object {
  if (!isObject(value)) return false
  const proto = Object.getPrototypeOf(value)
  return proto === Object.prototype || proto === null
}

/**
 * @ignore
 * @internal
 */
export function isObject(value: any): value is Object {
  return value !== null && typeof value === "object"
}

/**
 * @ignore
 * @internal
 */
export function isPrimitive(value: any): value is PrimitiveValue {
  switch (typeof value) {
    case "number":
    case "string":
    case "boolean":
    case "undefined":
    case "bigint":
      return true
  }
  return value === null
}

/**
 * @ignore
 * @internal
 */
export function debugFreeze(value: object) {
  if (inDevMode()) {
    Object.freeze(value)
  }
}

/**
 * @ignore
 * @internal
 */
export function deleteFromArray<T>(array: T[], value: T): boolean {
  let index = array.indexOf(value)
  if (index >= 0) {
    array.splice(index, 1)
    return true
  }
  return false
}

/**
 * @ignore
 * @internal
 */
export function isMap(val: any): val is Map<any, any> | ObservableMap {
  return val instanceof Map || isObservableMap(val)
}

/**
 * @ignore
 * @internal
 */
export function isSet(val: any): val is Set<any> | ObservableSet {
  return val instanceof Set || isObservableSet(val)
}

/**
 * @ignore
 * @internal
 */
export function isArray(val: any): val is any[] | IObservableArray {
  return Array.isArray(val) || isObservableArray(val)
}

/**
 * Checks whether an array is non-empty.
 *
 * @param arr Array to check.
 * @returns `true` if `arr` is non-empty, otherwise `false`.
 */
export function isNonEmptyArray<T>(arr: T[]): arr is NonEmptyArray<T> {
  return arr.length !== 0
}

/**
 * @ignore
 * @internal
 */
export function inDevMode(): boolean {
  return process.env.NODE_ENV !== "production"
}

/**
 * @ignore
 * @internal
 */
export function assertIsObject(value: unknown, argName: string): asserts value is object {
  if (!isObject(value)) {
    throw failure(`${argName} must be an object`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsPlainObject(value: unknown, argName: string): asserts value is object {
  if (!isPlainObject(value)) {
    throw failure(`${argName} must be a plain object`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsObservableObject(value: unknown, argName: string): asserts value is object {
  if (!isObservableObject(value)) {
    throw failure(`${argName} must be an observable object`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsObservableArray(
  value: unknown,
  argName: string
): asserts value is IObservableArray {
  if (!isObservableArray(value)) {
    throw failure(`${argName} must be an observable array`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsMap(value: unknown, argName: string): asserts value is Map<any, any> {
  if (!isMap(value)) {
    throw failure(`${argName} must be a map`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsSet(value: unknown, argName: string): asserts value is Set<any> {
  if (!isSet(value)) {
    throw failure(`${argName} must be a set`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsFunction(value: unknown, argName: string): asserts value is Function {
  if (typeof value !== "function") {
    throw failure(`${argName} must be a function`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsPrimitive(
  value: unknown,
  argName: string
): asserts value is PrimitiveValue {
  if (!isPrimitive(value)) {
    throw failure(`${argName} must be a primitive`)
  }
}

/**
 * @ignore
 * @internal
 */
export function assertIsString(value: unknown, argName: string): asserts value is string {
  if (typeof value !== "string") {
    throw failure(`${argName} must be a string`)
  }
}

/**
 * @ignore
 * @internal
 */
export interface DecorateMethodOrFieldData {
  target: any
  propertyKey: string
  baseDescriptor?: PropertyDescriptor & { initializer?: () => any }
}

const decoratorsSymbol = Symbol("decorators")

type WrapFunction = (data: DecorateMethodOrFieldData, fn: any) => any
type LateInitializationFunctionsArray = ((instance: any) => void)[]

/**
 * @ignore
 * @internal
 */
export function addLateInitializationFunction(target: any, fn: (instance: any) => void) {
  let decoratorsArray: LateInitializationFunctionsArray = target[decoratorsSymbol]
  if (!decoratorsArray) {
    decoratorsArray = []
    addHiddenProp(target, decoratorsSymbol, decoratorsArray)
  }
  decoratorsArray.push(fn)
}

/**
 * @ignore
 * @internal
 */
export function decorateWrapMethodOrField(
  decoratorName: string,
  data: DecorateMethodOrFieldData,
  wrap: WrapFunction
): any {
  const { target, propertyKey, baseDescriptor } = data

  const addFieldDecorator = () => {
    addLateInitializationFunction(target, (instance) => {
      instance[propertyKey] = wrap(data, instance[propertyKey])
    })
  }

  if (baseDescriptor) {
    if (baseDescriptor.get !== undefined) {
      throw failure(`@${decoratorName} cannot be used with getters`)
    }

    if (baseDescriptor.value) {
      // babel / typescript - method decorator
      // @action method() { }
      return {
        enumerable: false,
        writable: true,
        configurable: true,
        value: wrap(data, baseDescriptor.value),
      }
    } else {
      // babel - field decorator: @action method = () => {}
      addFieldDecorator()
    }
  } else {
    // typescript - field decorator
    addFieldDecorator()
  }
}

/**
 * @ignore
 * @internal
 */
export function runLateInitializationFunctions(instance: any): void {
  const fns: LateInitializationFunctionsArray | undefined = instance[decoratorsSymbol]
  if (fns) {
    for (const fn of fns) {
      fn(instance)
    }
  }
}

const warningsAlreadyDisplayed = new Set<string>()

/**
 * @ignore
 * @internal
 */
export function logWarning(type: "warn" | "error", msg: string, uniqueKey?: string): void {
  if (uniqueKey) {
    if (warningsAlreadyDisplayed.has(uniqueKey)) {
      return
    }
    warningsAlreadyDisplayed.add(uniqueKey)
  }

  msg = "[mobx-keystone] " + msg
  switch (type) {
    case "warn":
      console.warn(msg)
      break
    case "error":
      console.error(msg)
      break
    default:
      throw failure(`unknown log type - ${type}`)
  }
}

const notMemoized = Symbol("notMemoized")

/**
 * @ignore
 * @internal
 */
export function lateVal<TF extends (...args: any[]) => any>(getter: TF): TF {
  let memoized: TF | typeof notMemoized = notMemoized

  const fn = (...args: any[]): any => {
    if (memoized === notMemoized) {
      memoized = getter(...args)
    }
    return memoized
  }

  return fn as TF
}

/**
 * @ignore
 * @internal
 */
export function lazy<V>(valueGen: () => V): () => V {
  let inited = false
  let val: V | undefined

  return (): V => {
    if (!inited) {
      val = valueGen()
      inited = true
    }
    return val!
  }
}

// just to ensure import * is kept properly
/**
 * @ignore
 * @internal
 */
export const mobx6 = {
  // eslint-disable-next-line no-useless-concat
  makeObservable: (mobx as any)["makeObservable" + ""] as typeof mobx["makeObservable"],
}

/**
 * @ignore
 * @internal
 */
export function getMobxVersion(): number {
  if (mobx6.makeObservable!) {
    return 6
  } else {
    return 5
  }
}
