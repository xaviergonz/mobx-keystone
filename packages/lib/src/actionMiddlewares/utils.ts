import { modelIdKey } from "../model/metadata"
import { isModel } from "../model/utils"
import type { RootPath } from "../parent/path"
import type { Path } from "../parent/pathTypes"

/**
 * @internal
 */
export function rootPathToTargetPathIds(rootPath: RootPath<any>): (string | null)[] {
  const targetPathIds: (string | null)[] = []

  for (let i = 0; i < rootPath.path.length; i++) {
    const targetObj = rootPath.pathObjects[i + 1] // first is root, we don't care about its ID
    const targetObjId = isModel(targetObj) ? targetObj[modelIdKey] ?? null : null
    targetPathIds.push(targetObjId)
  }

  return targetPathIds
}

/**
 * @internal
 */
export function pathToTargetPathIds(root: any, path: Path): (string | null)[] {
  const targetPathIds: (string | null)[] = []
  let current = root // we don't care about the root ID

  for (let i = 0; i < path.length; i++) {
    current = current[path[i]]
    const targetObjId = isModel(current) ? current[modelIdKey] ?? null : null
    targetPathIds.push(targetObjId)
  }

  return targetPathIds
}
