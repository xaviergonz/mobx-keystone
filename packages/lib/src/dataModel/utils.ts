import type { ModelClass } from "../modelShared/BaseModelShared"
import { failure } from "../utils"
import type { AnyDataModel } from "./BaseDataModel"
import { _BaseDataModel } from "./_BaseDataModel"

/**
 * Checks if an object is a data model instance.
 *
 * @param model
 * @returns
 */
export function isDataModel(model: unknown): model is AnyDataModel {
  return model instanceof _BaseDataModel
}

/**
 * @ignore
 * @internal
 *
 * Asserts something is actually a data model.
 *
 * @param model
 * @param argName
 */
export function assertIsDataModel(
  model: unknown,
  argName: string,
  customErrMsg = "must be a data model instance"
): asserts model is AnyDataModel {
  if (!isDataModel(model)) {
    throw failure(`${argName} ${customErrMsg}`)
  }
}

/**
 * @ignore
 * @internal
 */
export function isDataModelClass(modelClass: unknown): modelClass is ModelClass<AnyDataModel> {
  if (typeof modelClass !== "function") {
    return false
  }

  if (modelClass !== _BaseDataModel && !(modelClass.prototype instanceof _BaseDataModel)) {
    return false
  }

  return true
}

/**
 * @ignore
 * @internal
 */
export function assertIsDataModelClass(
  modelClass: unknown,
  argName: string
): asserts modelClass is ModelClass<AnyDataModel> {
  if (typeof modelClass !== "function") {
    throw failure(`${argName} must be a class`)
  }

  if (modelClass !== _BaseDataModel && !(modelClass.prototype instanceof _BaseDataModel)) {
    throw failure(`${argName} must extend DataModel`)
  }
}
