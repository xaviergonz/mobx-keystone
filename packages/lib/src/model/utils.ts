import { SnapshotInOfModel } from "../snapshot/SnapshotOf"
import { failure, isPlainObject } from "../utils"
import { modelMetadataKey } from "./metadata"
import { AnyModel, BaseModel, ModelClass } from "./Model"

/**
 * Checks if an object is a model instance.
 *
 * @param model
 * @returns
 */
export function isModel(model: any): model is AnyModel {
  return model instanceof BaseModel
}

/**
 * @ignore
 *
 * Asserts something is actually a model.
 *
 * @param model
 * @param argName
 */
export function assertIsModel(
  model: AnyModel,
  argName: string,
  customErrMsg = "must be a model instance"
) {
  if (!isModel(model)) {
    throw failure(`${argName} ${customErrMsg}`)
  }
}

/**
 * @ignore
 */
export function isModelClass(modelClass: any): modelClass is ModelClass<AnyModel> {
  if (typeof modelClass !== "function") {
    return false
  }

  if (modelClass !== BaseModel && !(modelClass.prototype instanceof BaseModel)) {
    return false
  }

  return true
}

/**
 * @ignore
 */
export function assertIsModelClass(modelClass: ModelClass<AnyModel>, argName: string): void {
  if (typeof modelClass !== "function") {
    throw failure(`${argName} must be a class`)
  }

  if (modelClass !== BaseModel && !(modelClass.prototype instanceof BaseModel)) {
    throw failure(`${argName} must extend Model`)
  }
}

/**
 * @ignore
 */
export function isModelSnapshot(sn: any): sn is SnapshotInOfModel<AnyModel> {
  return isPlainObject(sn) && !!sn[modelMetadataKey]
}

/**
 * @ignore
 */
export function checkModelDecoratorArgs(fnName: string, target: any, propertyKey: string) {
  if (typeof propertyKey !== "string") {
    throw failure(fnName + " cannot be used over symbol properties")
  }

  const errMessage = fnName + " must be used over model classes or instances"

  if (!target) {
    throw failure(errMessage)
  }

  // check target is a model object or extended class
  if (
    !(target instanceof BaseModel) &&
    target !== BaseModel &&
    !(target.prototype instanceof BaseModel)
  ) {
    throw failure(errMessage)
  }
}
