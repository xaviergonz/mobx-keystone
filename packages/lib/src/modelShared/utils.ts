import type { AnyDataModel } from "../dataModel/BaseDataModel"
import { isDataModelClass } from "../dataModel/utils"
import type { AnyModel } from "../model/BaseModel"
import { isModelClass } from "../model/utils"
import { failure } from "../utils"
import type { ModelClass } from "./BaseModelShared"

/**
 * @internal
 *
 * Asserts something is actually a class or data model.
 *
 * @param model
 * @param argName
 */
export function assertIsClassOrDataModelClass(
  model: unknown,
  argName: string,
  customErrMsg = "must be a class or data model class"
): asserts model is ModelClass<AnyModel> | ModelClass<AnyDataModel> {
  if (!isModelClass(model) && !isDataModelClass(model)) {
    throw failure(`${argName} ${customErrMsg}`)
  }
}
