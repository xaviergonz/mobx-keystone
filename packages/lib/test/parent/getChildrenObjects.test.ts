import { computed, isObservableArray, reaction } from "mobx"
import { getChildrenObjects, model, Model, prop, runUnprotected } from "../../src"
import "../commonSetup"
import { autoDispose } from "../utils"

@model("C")
class C extends Model({ id: prop<string>(), children: prop<C[] | undefined>(undefined) }) {}

const deep = true

test(`getChildrenObjects {deep: ${deep}}`, () => {
  const getChildren = (n: object) => getChildrenObjects(n, { deep })

  const c = new C({
    id: "1",
    children: [
      new C({
        id: "1-1",
        children: [new C({ id: "1-1-1" }), new C({ id: "1-1-2" })],
      }),
      new C({
        id: "1-2",
        children: [new C({ id: "1-2-1" }), new C({ id: "1-2-2" })],
      }),
      new C({
        id: "1-3",
        children: [new C({ id: "1-3-1" }), new C({ id: "1-3-2" })],
      }),
    ],
  })

  const getChildrenIds = (n: any) =>
    [...getChildren(n).values()].map((n2: any) => (isObservableArray(n2) ? "array" : n2.id))

  let initialChildren = getChildren(c)
  let initialChildren1 = getChildren(c.children![0])
  let initialChildren3 = getChildren(c.children![2])

  let childrenChanges = 0
  let children1Changes = 0
  let children3Changes = 0

  autoDispose(
    reaction(
      () => getChildren(c),
      () => {
        childrenChanges++
      }
    )
  )
  autoDispose(
    reaction(
      () => getChildren(c.children![0]),
      () => {
        children1Changes++
      }
    )
  )
  const children3 = c.children![2]
  const computedChildren3 = computed(() => getChildren(children3))
  autoDispose(
    reaction(
      () => computedChildren3.get(),
      () => {
        children3Changes++
      }
    )
  )

  const expectChildrenChanges = (c: number, c1: number, c3: number) => {
    expect(childrenChanges).toBe(c)
    expect(children1Changes).toBe(c1)
    expect(children3Changes).toBe(c3)
    childrenChanges = 0
    children1Changes = 0
    children3Changes = 0
  }

  // should always return the same set when unchanged
  expect(getChildren(c)).toBe(initialChildren)
  expect(getChildren(c)).toBe(initialChildren)
  expect(getChildren(c.children![0])).toBe(initialChildren1)
  expect(getChildren(c.children![0])).toBe(initialChildren1)
  expect(getChildren(c.children![2])).toBe(initialChildren3)
  expect(getChildren(c.children![2])).toBe(initialChildren3)

  expect(getChildrenIds(c)).toMatchSnapshot("initial")
  expectChildrenChanges(0, 0, 0)

  // delete some
  let deleted!: C
  runUnprotected(() => {
    deleted = c.children!.pop()!
  })

  expect(getChildrenIds(c)).toMatchSnapshot("after delete")
  expectChildrenChanges(1, 0, 0)

  // add them back
  runUnprotected(() => {
    c.children!.push(deleted)
  })

  expect(getChildrenIds(c)).toMatchSnapshot("after re-add")
  expectChildrenChanges(1, 0, 0)
})
