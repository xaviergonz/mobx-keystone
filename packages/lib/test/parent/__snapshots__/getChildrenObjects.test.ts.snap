// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getChildrenObjects {deep: true}: after delete 1`] = `
Array [
  "array",
  "1-1",
  "array",
  "1-1-1",
  "1-1-2",
  "1-2",
  "array",
  "1-2-1",
  "1-2-2",
]
`;

exports[`getChildrenObjects {deep: true}: after re-add 1`] = `
Array [
  "array",
  "1-1",
  "array",
  "1-1-1",
  "1-1-2",
  "1-2",
  "array",
  "1-2-1",
  "1-2-2",
  "1-3",
  "array",
  "1-3-1",
  "1-3-2",
]
`;

exports[`getChildrenObjects {deep: true}: initial 1`] = `
Array [
  "array",
  "1-1",
  "array",
  "1-1-1",
  "1-1-2",
  "1-2",
  "array",
  "1-2-1",
  "1-2-2",
  "1-3",
  "array",
  "1-3-1",
  "1-3-2",
]
`;
